h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -147,7 +148,10 @@ public boolean stopEarly(Model model, ScoringInfo[] sk) {
      * Cached names of used hyper parameters.
      */
     final protected String[] _hyperParamNames;
+    
+    final protected String[] _hyperParamNamesConstraint;  // model parameters specified in constraints of hyper parameters
 
+    protected Map<String, Object>[] _hyperParamConstraints;
```

Comment:
if these are always lists (probably due to parser), let's at least force the type to `List` instead of `Object`, your code below will also become simpler.

But let's ask more: we want to have `Map<String, Object[]>[] _hyperParamConstraints;` here, same type as for `hyperParams`.
This would allow us to use the same logic everywhere we can use `hyperParams`: a typical example would be to support nested grouping later.
To enforce `Map<String, Object[]>[] _hyperParamConstraints`, this should be done in `GridSearchSchema#fillFromParms`: 
```
        // Convert lists and singletons into arrays
        for (Map.Entry<String, Object> e : m.entrySet()) {
          Object o = e.getValue();
          Object[] o2 = o instanceof List ? ((List) o).toArray() : new Object[]{o};

          hyper_parameters.put(e.getKey(),o2);
        }
```

This code above could be adapted to handle `CONSTRAINTS` key differently:
```
private static Map<String, Object[]> paramValuesToArray(Map<String, Object> params) {
        Map<String, Object[]> result = new HashMap<>();
        for (Map.Entry<String, Object> e : m.entrySet()) {
          String k = e.getKey();
          Object v = e.getValue();
          Object[] arr = CONSTRAINTS.equals(k) ? paramValuesToArray(v)
                           : v instanceof List ? ((List) v).toArray() 
                           : new Object[]{v};
          result.put(k, v);
        }
        return result;
```
Now we have arrays everywhere, and besides it handles nested constraints.

Reply:
Thank you for this suggestion - reduced code in the hyperspace walker and improved the design. Very elegant solution.